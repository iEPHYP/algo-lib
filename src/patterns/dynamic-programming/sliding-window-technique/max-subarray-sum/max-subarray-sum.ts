// S(i) = S(i-1) - a(i-1) + a(i + k - 1)
// S(0) = SUM(0, k);
// Complexity O(n)T, O(1)S
export const maxSubarraySum = (
  array: number[],
  seriesLength: number
): { sum: number | null; subarray: number[] } => {
  if (array.length < seriesLength || seriesLength <= 0) {
    return { sum: null, subarray: [] };
  }

  const firstSeries = array.slice(0, seriesLength);
  const firstSeriesSum = firstSeries.reduce((previousSum, item) => previousSum + item, 0);

  let maxSum: { sum: number; subarray: number[] } = { sum: firstSeriesSum, subarray: firstSeries };

  let previousSeriesSum = firstSeriesSum;
  for (let i = 1; i <= array.length - seriesLength; i++) {
    const nextSeries = array.slice(i, i + seriesLength);
    const nextSeriesSum = previousSeriesSum - array[i - 1] + array[i + seriesLength - 1];
    if (maxSum.sum < nextSeriesSum) {
      maxSum = { sum: nextSeriesSum, subarray: nextSeries };
    }
    previousSeriesSum = nextSeriesSum;
  }

  return maxSum;
};
