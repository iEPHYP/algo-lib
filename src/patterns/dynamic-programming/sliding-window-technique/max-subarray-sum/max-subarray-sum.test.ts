import { maxSubarraySum } from './max-subarray-sum';

test(maxSubarraySum.name + ' works properly', () => {
  expect(maxSubarraySum([1, 2, 5, 2, 8, 1, 5], 2)).toMatchSnapshot(
    'maxSubarraySum([1, 2, 5, 2, 8, 1, 5], 2)'
  );
  expect(maxSubarraySum([1, 2, 5, 2, 8, 1, 5], 4)).toMatchSnapshot(
    'maxSubarraySum([1, 2, 5, 2, 8, 1, 5], 4)'
  );
  expect(maxSubarraySum([4, 2, 1, 6], 1)).toMatchSnapshot('maxSubarraySum([4, 2, 1, 6], 1)');
  expect(maxSubarraySum([4, 2, 1, 6, 2], 4)).toMatchSnapshot('maxSubarraySum([4, 2, 1, 6, 2], 4)');
  expect(maxSubarraySum([], 4)).toMatchSnapshot('maxSubarraySum([], 4)');
  expect(maxSubarraySum([1, 5, 7], 4)).toMatchSnapshot('maxSubarraySum([1, 5, 7], 4)');
  expect(maxSubarraySum([1, 5, 7], 0)).toMatchSnapshot('maxSubarraySum([1, 5, 7], 0)');
});
