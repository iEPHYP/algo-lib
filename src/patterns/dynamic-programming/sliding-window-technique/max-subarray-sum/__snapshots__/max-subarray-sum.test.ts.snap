// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maxSubarraySum works properly: maxSubarraySum([], 4) 1`] = `
{
  "subarray": [],
  "sum": null,
}
`;

exports[`maxSubarraySum works properly: maxSubarraySum([1, 2, 5, 2, 8, 1, 5], 2) 1`] = `
{
  "subarray": [
    2,
    8,
  ],
  "sum": 10,
}
`;

exports[`maxSubarraySum works properly: maxSubarraySum([1, 2, 5, 2, 8, 1, 5], 4) 1`] = `
{
  "subarray": [
    2,
    5,
    2,
    8,
  ],
  "sum": 17,
}
`;

exports[`maxSubarraySum works properly: maxSubarraySum([1, 5, 7], 0) 1`] = `
{
  "subarray": [],
  "sum": null,
}
`;

exports[`maxSubarraySum works properly: maxSubarraySum([1, 5, 7], 4) 1`] = `
{
  "subarray": [],
  "sum": null,
}
`;

exports[`maxSubarraySum works properly: maxSubarraySum([4, 2, 1, 6, 2], 4) 1`] = `
{
  "subarray": [
    4,
    2,
    1,
    6,
  ],
  "sum": 13,
}
`;

exports[`maxSubarraySum works properly: maxSubarraySum([4, 2, 1, 6], 1) 1`] = `
{
  "subarray": [
    6,
  ],
  "sum": 6,
}
`;
